### ISSUE TITLE
Suggestions CLI

### ISSUE THREAD

**Issue by chartgerink** [hypergraph-xyz/cli#11]
:wave:

I was just working with the CLI and noticed a few things. These are not musts, but suggestions

- Create profile -> Display title → name? (to prevent confusion)
- Read metadata -> have the prompt select based on title?
- Update metadata -> have the prompt select based on title for writables?
- Old title dark grey when updating metadata, not readable on dark terminal?
- Capitalize names for consistency when updating metadata (now reads `title` not `Title` as in create)
- List writable modules -> Show the titles for selection?
- Add a whole `Open module` optionality? This would open `~/.p2pcommons/<hash>/`
- Select main file for module? This would show the files in the module, allows to select one. This would be a file explorer kind of selection, but might also be done in a popup?

👋

**Comment by juliangruber** [comment]
> * Create profile -> Display title → name? (to prevent confusion)

This is a bit confusing to implement, because a `profile` module only will have a `title` and not a `name` in the data structure underneath, however this suggests making it seem like it has a `name` and not a `title`. What about updating the space to remove `title` from `profile` modules and using `name` instead? `title` is confusing there in general and could be mistaken for the professional title.

> * Read metadata -> have the prompt select based on title?
> * Update metadata -> have the prompt select based on title for writables?

Do you mean that in addition to `hypergraph read <link>` we could support `hypergraph read <title>` in which case it would search for modules with that title and then output their data?

Should the prompt then only select on title, not hash? Or either (which is totally possible because we can easily tell apart a hash from any other string)?

> * Old title dark grey when updating metadata, not readable on dark terminal?

This is how it looks for me:

<img width="675" alt="Screenshot 2019-10-23 at 07 51 56" src="https://user-images.githubusercontent.com/10247/67361829-01614080-f56a-11e9-8a85-74acef6f2243.png">

Could it be that you have tweaked your color schema? With the library we're currently using we can't change the color of that default value (yet).

> * Capitalize names for consistency when updating metadata (now reads `title` not `Title` as in create)

Great suggestion, done!

> List writable modules -> Show the titles for selection?

Good idea, I would implement this once you confirm I understand your previous "search on title" suggestions correctly.

> * Add a whole `Open module` optionality? This would open `~/.p2pcommons/<hash>/`

Done! `hypergraph open <hash>`.

> * Select main file for module? This would show the files in the module, allows to select one. This would be a file explorer kind of selection, but might also be done in a popup?

This would be part of the `update` prompt, right?

**Comment by chartgerink** [comment]
> This is a bit confusing to implement, because a profile module only will have a title and not a name in the data structure underneath, however this suggests making it seem like it has a name and not a title. What about updating the space to remove title from profile modules and using name instead? title is confusing there in general and could be mistaken for the professional title.

Hm yea that is true. My main reason for using `title` instead of `name` for profile was because that is how the Dat archive metadata is called throughout the Dat universe as I've seen it. Beaker has had a tendency to not index variable names either. For a `type: profile` module the name is the title of the module, not the title of the person. At least, that was my thinking. All a bit confusing, I agree 🦐 

>Do you mean that in addition to hypergraph read <link> we could support hypergraph read <title> in which case it would search for modules with that title and then output their data?

Yea so that when you do `hypergraph read` it prompts with something like

```
Select module:

+ this is module 1
+ Theory of farts
+ Predictions about the sulphate contents of farts
```
as in [`prompts` select](https://github.com/terkelg/prompts#selectmessage-choices-initial-hint-warn)

> Should the prompt then only select on title, not hash? Or either (which is totally possible because we can easily tell apart a hash from any other string)?

I don't know? I think hashes are confusing to people because they're abstract, so we can move them to the back as much as possible. However, I can understand it makes sense to have the function for dev but just not have it in the CLI map? 

> Could it be that you have tweaked your color schema? With the library we're currently using we can't change the color of that default value (yet).

I have the default Linux mint schema, but this is not super important so let's just park that and only revisit it if people complain.

> Great suggestion, done!

😻 

> Good idea, I would implement this once you confirm I understand your previous "search on title" suggestions correctly.

👍 I was thinking als the [`prompts` select](https://github.com/terkelg/prompts#selectmessage-choices-initial-hint-warn) style options here

> Done! hypergraph open <hash>.

💯 

> This would be part of the update prompt, right?

Could be. That does bury it a bit, but then again, it is a logical step (once you know it). We can put it there first and see how intuitive it is upon usage?

**Comment by juliangruber** [comment]
> Hm yea that is true. My main reason for using `title` instead of `name` for profile was because that is how the Dat archive metadata is called throughout the Dat universe as I've seen it. Beaker has had a tendency to not index variable names either. For a `type: profile` module the name is the title of the module, not the title of the person. At least, that was my thinking. All a bit confusing, I agree 🦐

This is now done 🎆 It required a larger refactor but I'm happy with the results. It will be part of the next release :)



> Yea so that when you do `hypergraph read` it prompts with something like
> 
> ```
> Select module:
> 
> + this is module 1
> + Theory of farts
> + Predictions about the sulphate contents of farts
> ```
> 
> as in [`prompts` select](https://github.com/terkelg/prompts#selectmessage-choices-initial-hint-warn)

We currently don't enforce the title to be a non-empty string, which also the spec doesn't require. In that case, should we print the url if a module doesn't have one?

**Comment by chartgerink** [comment]
> This is now done :fireworks: It required a larger refactor but I'm happy with the results. It will be part of the next release :)

So now it displays name for `type: profile` where in the `dat.json` is `title`? :see_no_evil: Making sure for consistency and not to create conflicts with the sdk.

> We currently don't enforce the title to be a non-empty string, which also the spec doesn't require. In that case, should we print the url if a module doesn't have one?

I think we cleared that up with https://github.com/p2pcommons/specs/commit/b0fb3dee6b9babe1ca43f43b9736011b74c562e4

**Comment by juliangruber** [comment]
> So now it displays name for `type: profile` where in the `dat.json` is `title`? 🙈 Making sure for consistency and not to create conflicts with the sdk.

Yup that's exactly what it does :)

**Comment by juliangruber** [comment]
All done!

Released as 0.2.4
